tm/src/main/java/com/alibaba/fescar/tm/api/TransactionalTemplate.java

```diff
@@ -35,51 +40,93 @@ public Object execute(TransactionalExecutor business) throws TransactionalExecut
         // 1. get or create a transaction
         GlobalTransaction tx = GlobalTransactionContext.getCurrentOrCreate();
 
-        // 2. begin transaction
         try {
-            tx.begin(business.timeout(), business.name());
 
-        } catch (TransactionException txe) {
-            throw new TransactionalExecutor.ExecutionException(tx, txe,
-                TransactionalExecutor.Code.BeginFailure);
+            // 2. begin transaction
+            try {
+                triggerBeforeBegin();
+                tx.begin(business.timeout(), business.name());
+                triggerAfterBegin();
+            } catch (TransactionException txe) {
+                throw new TransactionalExecutor.ExecutionException(tx, txe,
+                        TransactionalExecutor.Code.BeginFailure);
 
-        }
+            }
+            Object rs = null;
+            try {
 
-        Object rs = null;
-        try {
+                // Do Your Business
+                rs = business.execute();
 
-            // Do Your Business
-            rs = business.execute();
+            } catch (Throwable ex) {
 
-        } catch (Throwable ex) {
+                // 3. any business exception, rollback.
+                try {
+                    tx.rollback();
 
-            // 3. any business exception, rollback.
-            try {
-                tx.rollback();
+                    // 3.1 Successfully rolled back
+                    throw new TransactionalExecutor.ExecutionException(tx, TransactionalExecutor.Code.RollbackDone, ex);
+
+                } catch (TransactionException txe) {
+                    // 3.2 Failed to rollback
+                    throw new TransactionalExecutor.ExecutionException(tx, txe,
+                            TransactionalExecutor.Code.RollbackFailure, ex);
 
-                // 3.1 Successfully rolled back
-                throw new TransactionalExecutor.ExecutionException(tx, TransactionalExecutor.Code.RollbackDone, ex);
+                }
+
+            }
+            // 4. everything is fine, commit.
+            try {
+                triggerBeforeCommit();
```

Comment:
Exception should be properly handled. I don't think users would like to fail committing by their hook.

Reply:
fixed it in next commit

Reply:
the hook help us to add more function in the transaction life cycle
 like 
  1.trace the transaction execute status
  2. customer logger  to help they find the bad transaction  
  3.some case we need to do something when the  commit all success